Adding docs for upgrade steps to 14.8 (#2699)
chef/chef-server: docs-chef-io/content/server/upgrades.md

Supported Release
: Chef Infra Server 13 and later are currently supported Chef Software releases. Earlier releases are no longer supported as of 12/31/2020. For more information about supported Chef Software see the [Supported Versions]({{< relref "/versions#supported-commercial-distributions" >}}) documentation.

## Release-Specific Steps

### Upgrading to 14.x

Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
{{% server_upgrade_duration %}}

The Chef Infra Server 14 upgrade does not automatically reindex existing external Elasticsearch installations.

#### Upgrading to 14.8

> Chef Infra Server 14.8 upgraded PostgreSQL from 9.6 to 13.3. The 14.8 upgrade process requires a one time downtime for vacuuming, upgrading, and re-indexing the database. The entire upgrade operation takes approximately 1 minute per 1000 nodes. This process may take longer depending on your server hardware and the size of the node objects in your Chef Infra Server.

Reviewer: could we add the size of the 1000 nodes for comparison?

{{< note >}}

> The attribute `private_chef['postgresql']['pg_upgrade_timeout']` in chef-server.rb is used to set the timeout value for pg upgrade. Set this value accordingly to the time required for your data size.

Reviewer: could we link to the options page here?

Supported Release
: Chef Infra Server 13 and later are currently supported Chef Software releases. Earlier releases are no longer supported as of 12/31/2020. For more information about supported Chef Software see the [Supported Versions]({{< relref "/versions#supported-commercial-distributions" >}}) documentation.

> ## Release Specific Steps

Reviewer 2: ```suggestion
## Release-Specific Steps
```

### Upgrading to 14.x

Chef Infra Server 14.0 moved from Solr to Elasticsearch as its search index.
{{% server_upgrade_duration %}}

The Chef Infra Server 14 upgrade does not automatically reindex existing external Elasticsearch installations.

#### Upgrading to 14.8

> Chef Infra Server 14.8 upgraded PostgreSQL from 9.6 to 13.3. The 14.8 upgrade process requires a one time downtime for vacuuming, upgrading and reindexing the database. The entire upgrade operation is estimated to take 1 minute per 1000 nodes. This process may take longer depending on your server hardware and the size of the node objects in your Chef Infra Server.

Reviewer 2: ```suggestion
Chef Infra Server 14.8 upgraded PostgreSQL from 9.6 to 13.3. The 14.8 upgrade process requires a one time downtime for vacuuming, upgrading, and re-indexing the database. The entire upgrade operation takes approximately 1 minute per 1000 nodes. This process may take longer depending on your server hardware and the size of the node objects in your Chef Infra Server.
```

##### Database Preparation

> 1. Running vacuum full on the PostgreSQL database is recommended if auto vacuuming is not set up. This process will reduce the size of the database by deleting unnecessary data and speeds up the migration. The vacuum full operation takes around 1 to 2 minutes per GB of data, depending on the complexity of the data. Additionally, the vacuum full operation will need an amount of free disk space at least as large as the size of your database.

Reviewer 2: ```suggestion
1. We recommend running `VACUUM FULL` on the PostgreSQL database if you don't have automatic vacuuming set up. This process will reduce the size of the database by deleting unnecessary data and speeds up the migration. The `VACUUM FULL` operation takes around 1 to 2 minutes per gigabyte of data depending on the complexity of the data. Additionally, the `VACUUM FULL` operation requires additional free disk space at least as large as the size of your database.
```

> 1. Running vacuum full on the PostgreSQL database is recommended if auto vacuuming is not set up. This process will reduce the size of the database by deleting unnecessary data and speeds up migration. The vacuum full operation takes around 1 to 2 minutes per GB of data, depending on the complexity of the data. Additionally, the vacuum full operation will need an amount of free disk space at least as large as the size of your database.

Reviewer: can we include some context about the number of nodes from out testing?

Reviewer: We also need to mention the additional diskspace required to run the vacuum full

    ```bash

>     /usr/lib/postgresql/9.6/bin/vaccumedb --all --full

Reviewer: is this path the same for all the installs? standalone, tiered and external postgres?

Reviewer 3: I don't think this is the right path

Reviewer 3: /opt/opscode/embedded/bin/vacuumdb

    ```

> 2. To allow for a full restoration to a previous release in the event of a failure we recommend backing up the PostgreSQL database before upgrading. See [Backup and Restore]({{< relref "server_backup_restore" >}}) for additional information.

Reviewer 2: ```suggestion
1. To allow for a full restoration to a previous release in the event of a failure, we recommend backing up the PostgreSQL database before upgrading. See [Backup and Restore]({{< relref "server_backup_restore" >}}) for additional information.
```

##### Upgrade Steps

Follow the Chef Infra Server upgrade instructions below.

##### Post-upgrade Steps

> Reindex the Chef Infra Server

Reviewer 2: ```suggestion
Reindex the Chef Infra Server:
```

> ##### Post-upgrade Steps

Reviewer: should the postupgrade section be after the upgrade steps?

Reviewer: And we can include a link to the postupgrade section at the end of install instructions.

Reviewer 3: I added a note that this proces is one time

Author: I added the link to post upgrade at the end of upgrade instruction in a note

> 1. Reindex the Chef Infra Server

Reviewer: we also need to specify that this is a one time operation for upgrading to any version >=14.8 the first time.

```bash
/opt/opscode/embedded/bin/reindexdb --all
```

{{< note >}}

> Estimates were performed on an 8 core 32 GB memory (t3.2xlarge) AWS EC2 instance with 2 orgs, 3 users having 565000+ nodes, 13000+ cookbooks, 29000+ databags, 274000+ envs and 281000+ roles. 

Reviewer 2: ```suggestion
Estimates were performed on an 8-core 32 GB memory (t3.2xlarge) AWS EC2 instance with 2 organizations, 3 users having 565,000+ nodes, 13,000+ cookbooks, 29,000+ databags, 274,000+ environments, and 281,000+ roles. 
```

The Chef Infra Server 14 upgrade does not automatically reindex existing external Elasticsearch installations.

> ### Upgrading to 14.8.x

Reviewer 3: ```suggestion
### Upgrading to 14.8
```

> Chef Infra Server 14.8 upgraded postgresql from 9.6 to 13.3. There are a couple of extra steps that need to be considered depending on the scenario. 

Reviewer: Can we add information here about what the expected downtime will be? And also point them to additional system requirements before getting started with the upgrade.

#### Recommendations

> 1. Running Vacuum full is recommended if auto vacuuming is not set up, as this will reduce the size of the db by deleting the unwanted data. We estimate the vacuum full operation will take about 84 minutes for 82 GB of data and it will also depend on the nature of the data also.

Reviewer: could we give ball-parks or multiplication factor here?

> 2. Taking a chef-srever-ctl backup or knife-ec-backup is recommended before starting the upgrade. The backup takes around 4 to 5 minutes per GB of data.

Reviewer: s/chef-srever-ctl/chef-server-ctl

The Chef Infra Server 14 upgrade does not automatically reindex existing external Elasticsearch installations.

### Upgrading to 14.8.x

> Chef Infra Server 14.8 upgraded PostgreSQL from 9.6 to 13.3. The 14.8 upgrade process requires downtime for vacuuming, upgrading and reindexing the database. There are extra steps that need to be considered when upgrading from versions less than 14.8 to versions greater than 14.8. Using current recommendations, the entire upgrade operation could take 1 minute per 1000 nodes or longer, depending on your server hardware and the complexity of your Chef data.

Reviewer 3: ```suggestion
Chef Infra Server 14.8 upgraded PostgreSQL from 9.6 to 13.3. The 14.8 upgrade process requires downtime for vacuuming, upgrading and reindexing the database. There are extra steps that need to be considered when upgrading from versions less than 14.8 to version 14.8 or later. Using current recommendations, the entire upgrade operation is estimated to take 1 minute per 1000 nodes. This process may take longer depending on your server hardware and the size of the node objects in your Chef Infra Server.
```

#### Pre-upgrade Recommendations

1. Running vacuum full is recommended if auto vacuuming is not set up, as this will reduce the size of the database by deleting unnecessary data. The vacuum full operation could take around 1 to 2 minutes per GB of data, depending on the nature of the data. Additionally, the vacuum full operation will need an amount of free disk space at least as large as the size of your database.

    ```bash
    /usr/lib/postgresql/9.6/bin/vaccumedb --all --full
    ```

2. Performing `chef-server-ctl backup` or `knife-ec-backup` is recommended before starting the upgrade. The backup takes around 4 to 5 minutes per GB of data.

#### Upgrade Steps

1. Follow the same instructions for upgrading to 14.

#### Post-upgrade steps

> 1. Reindex the Chef Infra server

Reviewer 3: ```suggestion
1. Reindex the Chef Infra Server
```

    ```bash
    /usr/lib/postgresql/13.3/bin/reindexdb --all
    ```

#### Additional Information

> 1. Estimates were performed on an 8 core 32 GB memory (t3.2xlarge EC2) machine.

Reviewer 3: ```suggestion
1. Estimates were performed on an 8 core 32 GB memory (t3.2xlarge) AWS EC2 instance.
```

> 2. The down time for standalone scenario upgrade with 158G of unvacuumed data was about 5hr 15min.

Reviewer 3: We need to make sure people understand how large this install was. We can get the node count + count for other objects like roles/envs/data bags/policyfiles/orgs/users. Otherwise everyone is going to read that and go "oh no I need 5 hours"

Author: I have added that along with the note

#### Upgrading to 14.8

> Chef Infra Server 14.8 upgrades PostgreSQL from 9.6 to 13.3. The 14.8 upgrade process requires a one-time downtime to vacuum, upgrade, and re-index the database. The entire upgrade operation takes approximately one minute per 1000 nodes (1000 nodes will approximately be 286MB). This process may take longer depending on your server hardware and the size of the node objects on your Chef Infra Server.

Reviewer 2: ```suggestion
Chef Infra Server 14.8 upgrades PostgreSQL from 9.6 to 13.3. The 14.8 upgrade process requires a one-time downtime to vacuum, upgrade, and re-index the database. The entire upgrade operation takes approximately one minute per 1000 nodes (1000 nodes is approximately 286MB). This process may take longer depending on your server hardware and the size of the node objects on your Chef Infra Server.
```

{{< note >}}

> The attribute `postgresql['pg_upgrade_timeout']` in [chef-server.rb]({{< relref "config_rb_server_optional_settings" >}}) is used to set the timeout value for pg upgrade. Set this value accordingly to the time required for your data size. 

Reviewer 2: ```suggestion
Set the `postgresql['pg_upgrade_timeout']` attribute in [chef-server.rb]({{< relref "config_rb_server_optional_settings" >}}) to the timeout value for the upgrade. Set this value based on the size of your data, where it take approximately one minute per 1,000 nodes which is approximately 286MB.
```

> The attribute `postgresql['pg_upgrade_timeout']` in [chef-server.rb]({{< relref "config_rb_server_optional_settings" >}}) is used to set the timeout value for pg upgrade. Set this value accordingly to the time required for your data size. 

Reviewer 2: Should `postgresql['pg_upgrade_timeout']` be added to config_rb_server_optional_settings.md page? Cause I don't see it in there.

Author: Ya, that is needed. I have updated the docs. Thanks!

    chef-server-ctl cleanup
    ```

{{< note >}}

> Check [post upgrade steps](#post-upgrade-steps) if you are upgrading from version less than 14.8 to version greater than or equal to 14.8

Reviewer 2: ```suggestion
Check the [post upgrade steps](#post-upgrade-steps) if you are upgrading from a version before Chef Infra Server 14.8 to a version greater than or equal to 14.8.
```

   chef-server-ctl cleanup
   ```

{{< note >}}

> Check [post upgrade steps](#post-upgrade-steps) if you are upgrading from version less than 14.8 to version greater than or equal to 14.8

Reviewer 2: ```suggestion
Check the [post upgrade steps](#post-upgrade-steps) if you are upgrading from a version before Chef Infra Server 14.8 to a version greater than or equal to 14.8.
```

#### Use Local Packages

{{% ctl_chef_server_install_features_manual %}}

> 

Reviewer 3: ```suggestion
```

> 

Reviewer 3: ```suggestion
```
